File: main/src/main/scala/org/clulab/dynet/ForwardLayer.scala
package org.clulab.dynet

import edu.cmu.dynet.{Dim, Expression, ExpressionVector, LookupParameter, Parameter, ParameterCollection}
import org.slf4j.{Logger, LoggerFactory}
import org.clulab.dynet.Utils.{ByLineIntBuilder, fromIndexToString, mkTransitionMatrix}
import org.clulab.scala.BufferedIterator
import org.clulab.scala.WrappedArrayBuffer._
import org.clulab.struct.Counter
import org.clulab.utils.Configured

import ForwardLayer._

abstract class ForwardLayer (val parameters:ParameterCollection,
                             val inputSize: Int,
                             val isDual: Boolean,
                             val t2i: Map[String, Int],
                             val i2t: Array[String],
                             val H: Parameter,
                             val rootParam: Parameter,
                             val distanceEmbeddingSize: Int,
                             val distanceLookupParameters: Option[LookupParameter],
                             val nonlinearity: Int,
                             val dropoutProb: Float)
  extends FinalLayer {

  def forward(inputExpressions: ExpressionVector,
              modifierHeadPairs: Option[IndexedSeq[ModifierHeadPair]],
              doDropout: Boolean): ExpressionVector = {
    val pH = Expression.parameter(H)
    val pRoot = Expression.parameter(rootParam)
    val emissionScores = new ExpressionVector()

    if(! isDual) {
      //
      // basic task
      //

      for (i <- inputExpressions.indices) {
        // the hidden state for the argument token
        var argExp = inputExpressions(i)

        // dropout on the hidden state
        argExp = Utils.expressionDropout(argExp, dropoutProb, doDropout)

        // forward layer + dropout on that
        var l1 = Utils.expressionDropout(pH * argExp, dropoutProb, doDropout)

        // the last nonlinearity
        l1 = nonlinearity match {
          case NONLIN_TANH => Expression.tanh(l1)
          case NONLIN_RELU => Expression.rectify(l1)
          case _ => l1
        }

        emissionScores.add(l1)
      }
    }

    else {
      //
      // dual task
      //

      if (modifierHeadPairs.isEmpty) {
        throw new RuntimeException("ERROR: dual task without information about head positions!")
      }

      for (pair <- modifierHeadPairs.get) {
        val modPosition = pair.modifier
        val headPosition = pair.head

        val argExp = Utils.expressionDropout(inputExpressions(modPosition), dropoutProb, doDropout)
        val predExp = if (headPosition >= 0) {
          // there is an explicit head in the sentence
          Utils.expressionDropout(inputExpressions(headPosition), dropoutProb, doDropout)
        } else {
          // the head is root. we use a dedicated Parameter for root
          Utils.expressionDropout(pRoot, dropoutProb, doDropout)
        }

        val ss =
          if (distanceLookupParameters.nonEmpty) {
            val distEmbedding = mkRelativePositionEmbedding(modPosition, headPosition)
            Expression.concatenate(argExp, predExp, distEmbedding)
          } else {
            Expression.concatenate(argExp, predExp)
          }

        var l1 = Utils.expressionDropout(pH * ss, dropoutProb, doDropout)

        l1 = nonlinearity match {
          case NONLIN_TANH => Expression.tanh(l1)
          case NONLIN_RELU => Expression.rectify(l1)
          case _ => l1 // nothing to do otherwise
        }

        emissionScores.add(l1)
      }
    }

    emissionScores
  }

  protected def mkRelativePositionEmbedding(modifier: Int, head: Int): Expression = {
    val dist =
      if(head == -1) { // root
        0
      } else { // head is an actual token
        //println(s" head $head modifier $modifier")
        val fullDist = head - modifier
        if(fullDist < -50) -50
        else if(fullDist > 50) 50
        else fullDist
      } + 50
    //println(s"Finding embedding for position $dist")
    Expression.lookup(distanceLookupParameters.get, dist)
  }

  override def inDim: Int = inputSize

  override def outDim: Int = t2i.size
}

object ForwardLayer {
  val logger: Logger = LoggerFactory.getLogger(classOf[ViterbiForwardLayer])

  val DEFAULT_DROPOUT_PROB: Float = Utils.DEFAULT_DROPOUT_PROBABILITY

  val TYPE_VITERBI = 1
  val TYPE_GREEDY = 2

  val NONLIN_NONE = 0
  val NONLIN_RELU = 1
  val NONLIN_TANH = 2

  val TYPE_GREEDY_STRING = "greedy"
  val TYPE_VITERBI_STRING = "viterbi"

  val DEFAULT_IS_DUAL = 0

  def load(parameters: ParameterCollection,
           x2iIterator: BufferedIterator[String]): ForwardLayer = {
    val inferenceType = new ByLineIntBuilder().build(x2iIterator, "inferenceType")

    inferenceType match {
      case TYPE_VITERBI => ViterbiForwardLayer.load(parameters, x2iIterator)
      case TYPE_GREEDY => GreedyForwardLayer.load(parameters, x2iIterator)
      case _ => throw new RuntimeException(s"ERROR: unknown forward layer type $inferenceType!")
    }
  }

  def initialize(config: Configured,
                 paramPrefix: String,
                 parameters: ParameterCollection,
                 labelCounter: Counter[String],
                 isDual: Boolean,
                 inputSize: Int): Option[ForwardLayer] = {
    if (!config.contains(paramPrefix)) {
      return None
    }

    val inferenceType = config.getArgString(paramPrefix + ".inference", Some("greedy"))
    val dropoutProb = config.getArgFloat(paramPrefix + ".dropoutProb", Some(ForwardLayer.DEFAULT_DROPOUT_PROB))

    val nonlinAsString = config.getArgString(paramPrefix + ".nonlinearity", Some(""))
    val nonlin = nonlinAsString match {
      case "relu" => NONLIN_RELU
      case "tanh" => NONLIN_TANH
      case "" => NONLIN_NONE
      case _ => throw new RuntimeException(s"ERROR: unknown non-linearity $nonlinAsString!")
    }

    val t2i = labelCounter.keySet.toList.sorted.zipWithIndex.toMap
    val i2t = fromIndexToString(t2i)

    val distanceEmbeddingSize = config.getArgInt(paramPrefix + ".distanceEmbeddingSize", Some(0))
    val distanceLookupParameters =
      if(distanceEmbeddingSize > 0) Some(parameters.addLookupParameters(101, Dim(distanceEmbeddingSize)))
      else None

    val actualInputSize = if(isDual) 2 * inputSize + distanceEmbeddingSize else inputSize

    val H = parameters.addParameters(Dim(t2i.size, actualInputSize))
    val rootParam = parameters.addParameters(Dim(inputSize))

    inferenceType match {
      case TYPE_GREEDY_STRING =>
        Some(new GreedyForwardLayer(parameters,
          inputSize, isDual,
          t2i, i2t, H, rootParam,
          distanceEmbeddingSize, distanceLookupParameters,
          nonlin, dropoutProb))
      case TYPE_VITERBI_STRING =>
        val T = mkTransitionMatrix(parameters, t2i)
        val layer = new ViterbiForwardLayer(parameters,
          inputSize, isDual,
          t2i, i2t, H, T, rootParam,
          distanceEmbeddingSize, distanceLookupParameters,
          nonlin, dropoutProb)
        layer.initializeTransitions()
        Some(layer)
      case _ =>
        new RuntimeException(s"ERROR: unknown inference type $inferenceType!")
        None
    }
  }
}

File: main/src/main/python/pytorch/forwardLayer.py
import torch
import torch.nn as nn
from torch.autograd import Variable
import torch.nn.functional as F

from pytorch.finalLayer import FinalLayer

from pytorch.utils import *

class ForwardLayer(FinalLayer):
    def __init__(self, inputSize, isDual, t2i, i2t, actualInputSize, nonlinearity, dropoutProb, spans = None):
        super().__init__()
        self.inputSize = inputSize
        self.isDual = isDual
        self.t2i = t2i
        self.i2t = i2t
        self.spans = spans
        self.nonlinearity = nonlinearity

        self.pH = nn.Linear(actualInputSize, len(t2i))
        nn.init.xavier_uniform_(self.pH.weight)
        self.pRoot = Variable(torch.rand(inputSize)) #TODO: Not sure about the shape here
        self.dropout = nn.Dropout(dropoutProb)

        self.inDim = spanLength(spans) if spans is not None else inputSize
        self.outDim = len(t2i)

    # remove pick span part to simplify the ONNX converting
    # def pickSpan(self, v, i):
    #     if self.spans is None:
    #         return v
    #     else:
    #         # Zheng: Will spans overlap?
    #         vs = list()
    #         for span in self.spans:
    #             e = torch.index_select(v, i, torch.tensor(range(span[0], span[1])))
    #             vs.append(e)
    #         return torch.cat(vs, dim=i)

    def forward(self, inputExpressions, headPositionsOpt = None):
        if not self.isDual:
            # Zheng: Why the for loop here? Can we just use matrix manipulation?
            argExp = self.dropout(inputExpressions)
            emissionScores = self.dropout(self.pH(argExp))
            if self.nonlinearity == NONLIN_TANH:
                emissionScores = F.tanh(emissionScores)
            elif self.nonlinearity == NONLIN_RELU:
                emissionScores = F.relu(emissionScores)
        else:
            emissionScores = list()
            if headPositionsOpt is None:
                raise RuntimeError("ERROR: dual task without information about head positions!")
            for i, e in enumerate(inputExpressions):
                headPosition = headPositionsOpt[i]
                argExp = self.dropout(e)
                if headPosition >= 0:
                    # there is an explicit head in the sentence
                    predExp = self.dropout(inputExpressions[headPosition])
                else:
                    # the head is root. we used a dedicated Parameter for root
                    predExp = self.dropout(self.pRoot)
                ss = torch.cat([argExp, predExp])
                l1 = self.dropout(self.pH(ss))
                if self.nonlinearity == NONLIN_TANH:
                    l1 = F.tanh(l1)
                elif self.nonlinearity == NONLIN_RELU:
                    l1 = F.relu(l1)
                emissionScores.append(l1)
            emissionScores = torch.stack(emissionScores)
        return emissionScores

    @staticmethod
    def load(x2i):
        from pytorch.greedyForwardLayer import GreedyForwardLayer
        from pytorch.viterbiForwardLayer import ViterbiForwardLayer
        inferenceType = x2i["inferenceType"]
        if inferenceType == TYPE_VITERBI or inferenceType == TYPE_VITERBI_STRING:#this is a temporary solution to handle a typo in viterbi forward layer...
            return ViterbiForwardLayer.load(x2i)
        elif inferenceType == TYPE_GREEDY or inferenceType == TYPE_GREEDY_STRING:
            return GreedyForwardLayer.load(x2i)
        else:
            raise RuntimeError(f"ERROR: unknown forward layer type {inferenceType}!")

    @staticmethod
    def initialize(config, paramPrefix, labelCounter, isDual, inputSize):
        from pytorch.greedyForwardLayer import GreedyForwardLayer
        from pytorch.viterbiForwardLayer import ViterbiForwardLayer
        if(not config.contains(paramPrefix)):
            return None

        inferenceType = config.get_string(paramPrefix + ".inference", "greedy")
        dropoutProb = config.get_float(paramPrefix + ".dropoutProb", DEFAULT_DROPOUT_PROBABILITY)

        nonlinAsString = config.get_string(paramPrefix + ".nonlinearity", "")
        if nonlinAsString in nonlin_map:
            nonlin = nonlin_map[nonlinAsString]
        else:
            raise RuntimeError(f"ERROR: unknown non-linearity {nonlinAsString}!")

        t2i = {t:i for i, t in enumerate(labelCounter.keys())}
        i2t = {i:t for t, i in t2i.items()}

        spanConfig = config.get_string(paramPrefix + ".span", "")
        if spanConfig is "":
            span = None
        else:
            span = parseSpan(spanConfig)

        if span:
            l = spanLength(span)
            actualInputSize = 2*l if isDual else l
        else:
            actualInputSize = 2*inputSize if isDual else inputSize

        if inferenceType == TYPE_GREEDY_STRING:
            return GreedyForwardLayer(inputSize, isDual, t2i, i2t, actualInputSize, nonlin, dropoutProb, span)
        elif inferenceType == TYPE_VITERBI_STRING:
            layer = ViterbiForwardLayer(inputSize, isDual, t2i, i2t, actualInputSize, nonlin, dropoutProb, span)
            return layer
        else:
            raise RuntimeError(f"ERROR: unknown inference type {inferenceType}!")
    
def spanLength(spans):
    return sum(end - start for start, end in spans)

def parseSpan(spanParam, inputSize=None):
    # Zheng: Why do we need inputSize here?
    spans = list()
    spanParamTokens = spanParam.split(",")
    for spanParamToken in spanParamTokens:
        # spanTokens = spanParamToken.split('-')
        # assert(len(spanTokens) == 2)
        # spans.append((int(spanTokens[0]), int(spanTokens[1])))
        token1, token2 = map(int, spanParamToken.split('-'))
        spans.append((token1, token2))
    return spans

def spanToString(spans):
    return ','.join(f'{start}-{end}' for start, end in spans)


























